core/search/boolean_filter.cpp

```diff
@@ -387,14 +382,14 @@ filter::prepared::ptr boolean_filter::prepare(
   // determine incl/excl parts
   std::vector<const filter*> incl;
   std::vector<const filter*> excl;
+  filters_t aux_filters; // filter added by optimization steps
 
   group_filters(incl, excl);
-  remove_excess(incl, excl, boost);
+  optimize(incl, excl, ord, boost, aux_filters);
```

Comment:
virtual `optimize` seems to become kludgy, e.g. now we have to instantiation use one more vector and extra heap allocation for auxiliary `all` filter, let's move optimization logic to corresponding `prepare` methods